nullaway/src/main/java/com/uber/nullaway/NullAway.java

```diff
@@ -694,14 +713,60 @@ private boolean fieldInitializedByPreviousInitializer(
     }
     // all the initializer blocks have run before any code inside a constructor
     constructors.stream().forEach((c) -> builder.putAll(c, initThusFar));
+    Symbol.ClassSymbol classSymbol = ASTHelpers.getSymbol(enclosingClass);
+    FieldInitEntities entities = class2Entities.get(classSymbol);
+    if (entities.instanceInitializerMethods().size() == 1) {
+      MethodTree initMethod = entities.instanceInitializerMethods().iterator().next();
+      // collect the fields that may not be initialized by *some* constructor NC
+      Set<Symbol> constructorUninitSymbols = class2ConstructorUninit.get(classSymbol);
+      // fields initialized after constructors is initThusFar + (nonNullFields - constructorUninit)
+      Sets.SetView<Element> initAfterConstructors =
+          Sets.union(
+              initThusFar,
+              Sets.difference(entities.nonnullInstanceFields(), constructorUninitSymbols));
+      builder.putAll(initMethod, initAfterConstructors);
+    }
+    if (entities.staticInitializerMethods().size() == 1) {
+      MethodTree staticInitMethod = entities.staticInitializerMethods().iterator().next();
+      // constructors aren't relevant here; just use initThusFar
+      builder.putAll(staticInitMethod, initThusFar);
+    }
     return builder.build();
   }
 
-  private boolean lhsOfAssignment(TreePath path) {
+  /**
+   * @param path tree path to read operation
+   * @return true if it is permissible to perform this read before the field has been initialized,
+   *     false otherwise
+   */
+  private boolean okToReadBeforeInitialized(TreePath path) {
     TreePath parentPath = path.getParentPath();
-    if (parentPath.getLeaf() instanceof AssignmentTree) {
-      AssignmentTree assignment = (AssignmentTree) parentPath.getLeaf();
-      return assignment.getVariable().equals(path.getLeaf());
+    Tree leaf = path.getLeaf();
+    Tree parent = parentPath.getLeaf();
+    if (parent instanceof AssignmentTree) {
+      // ok if it's actually a write
+      AssignmentTree assignment = (AssignmentTree) parent;
+      return assignment.getVariable().equals(leaf);
+    } else if (parent instanceof BinaryTree) {
+      // ok if we're comparing to null
+      BinaryTree binaryTree = (BinaryTree) parent;
+      Tree.Kind kind = binaryTree.getKind();
+      if (kind.equals(Tree.Kind.EQUAL_TO) || kind.equals(Tree.Kind.NOT_EQUAL_TO)) {
+        ExpressionTree left = binaryTree.getLeftOperand();
+        ExpressionTree right = binaryTree.getRightOperand();
+        return (left.equals(leaf) && right.getKind().equals(Tree.Kind.NULL_LITERAL))
+            || (right.equals(leaf) && left.getKind().equals(Tree.Kind.NULL_LITERAL));
+      }
+    } else if (parent instanceof MethodInvocationTree) {
+      // ok if it's invoking castToNonNull and the read is the argument
```

Comment:
How about passing it to other methods that take `@Nullable`? Do we care to allow that case?

Reply:
(I don't mean in this PR, just a possible extension)

Reply:
Yeah, the "right" thing to do here is to treat the field as `@Nullable` until it gets initialized.  That is a deeper change (to dataflow analysis, etc.), but probably the "right" way to do things.  For now I think we're ok with a narrow set of exclusions